pageserver/src/tenant/mgr.rs

```diff
@@ -1101,101 +1305,435 @@ pub(crate) async fn attach_tenant(
+                Err(TenantSlotUpsertError::InternalError(
+                    "Unexpected contents of TenantSlot".into(),
+                ))
+            }
+        }
+    }
+
+    /// Replace the InProgress slot with whatever was in the guard when we started
+    fn revert(mut self) {
+        if let Some(value) = self.old_value.take() {
+            match self.upsert(value) {
+                Err(TenantSlotUpsertError::InternalError(_)) => {
+                    // We already logged the error, nothing else we can do.
+                }
+                Err(TenantSlotUpsertError::MapState(_)) => {
+                    // If the map is shutting down, we need not replace anything
+                }
+                Ok(()) => {}
+            }
+        }
+    }
+
+    /// We may never drop our old value until it is cleanly shut down: otherwise we might leave
+    /// rogue background tasks that would write to the local tenant directory that this guard
+    /// is responsible for protecting
+    fn old_value_is_shutdown(&self) -> bool {
+        match self.old_value.as_ref() {
+            Some(TenantSlot::Attached(tenant)) => {
+                // TODO: PR #5711 will add a gate that enables properly checking that
+                // shutdown completed.
+                matches!(
+                    tenant.current_state(),
+                    TenantState::Stopping { .. } | TenantState::Broken { .. }
+                )
+            }
+            Some(TenantSlot::Secondary) => {
+                // TODO: when adding secondary mode tenants, this will check for shutdown
+                // in the same way that we do for `Tenant` above
+                true
+            }
+            Some(TenantSlot::InProgress(_)) => {
+                // A SlotGuard cannot be constructed for a slot that was already InProgress
+                unreachable!()
+            }
+            None => true,
+        }
+    }
+
+    /// The guard holder is done with the old value of the slot: they are obliged to already
+    /// shut it down before we reach this point.
+    fn drop_old_value(&mut self) -> Result<(), TenantSlotDropError> {
+        if !self.old_value_is_shutdown() {
+            Err(TenantSlotDropError::NotShutdown)
+        } else {
+            self.old_value.take();
+            Ok(())
+        }
     }
 }
 
-async fn tenant_map_upsert_slot<'a, F, R>(
-    tenant_id: TenantId,
-    upsert_fn: F,
-) -> Result<(), TenantMapInsertError>
-where
-    F: FnOnce(Option<TenantSlot>) -> R,
-    R: std::future::Future<Output = anyhow::Result<TenantSlot>>,
-{
-    let mut guard = TENANTS.write().await;
-    let m = match &mut *guard {
-        TenantsMap::Initializing => return Err(TenantMapInsertError::StillInitializing),
-        TenantsMap::ShuttingDown(_) => return Err(TenantMapInsertError::ShuttingDown),
+impl Drop for SlotGuard {
+    fn drop(&mut self) {
+        if !self.upserted {
```

Comment:
could you instead make `if self.upserted { return}`?

Reply:
b09891e52